### PR TITLE
flex_gemm: support block-scaled GEMMs through QUACK

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #193197
* #196183
* #192894
* #192664
* #192663
* #196321
* #196318
* #196320
* #196319
* __->__ #192839

### Human Note

With this we can finally do full lowp mlp (dense) in 
```Py

import torch
import torch.nn.functional as F
from torch._higher_order_ops import flex_gemm
from torch._higher_order_ops.flex_gemm import to_blocked
from torch._higher_order_ops.inline_asm_elementwise import inline_asm_elementwise


SCALED_MM = dict(
    scale_recipe_a=F.ScalingType.BlockWise1x32,
    scale_recipe_b=F.ScalingType.BlockWise1x32,
    swizzle_a=F.SwizzleType.SWIZZLE_32_4_4,
    swizzle_b=F.SwizzleType.SWIZZLE_32_4_4,
    output_dtype=torch.float32,
)


def quantize_mxfp8(x):
    """Illustrative local scales only (no tensorwise calibration/global scale)."""
    blocks = x.float().view(x.shape[0], -1, 32)
    # Round the block scale upward to a representable E8M0 power of two.
    scales = inline_asm_elementwise(
        blocks.abs().amax(-1, keepdim=True) / 448.0,
        asm_str="cvt.rp.satfinite.ue8m0x2.f32 $0, 0.0, $1;",
        constraints="=h,r",
        dtype=torch.float8_e8m0fnu,
    )
    normalized = (blocks * scales.float().reciprocal()).view_as(x)
    values = normalized.clamp(-448, 448).to(torch.float8_e4m3fn)
    return values, to_blocked(scales.squeeze(-1))


def silu_and_quantize(acc):
    # FP32 accumulator -> SiLU -> FP8, without a BF16 intermediate rounding.
    return quantize_mxfp8(F.silu(acc.float()))


def mlp(x, sx, w1, sw1, w2, sw2, *, tuned=True):
    options = {"backend": "QUACK", "tuned": tuned}
    hidden, sh = flex_gemm(
        F.scaled_mm,
        (x, w1.t(), sx, sw1),
        silu_and_quantize,
        gemm_kwargs=SCALED_MM,
        kernel_options=options,
    )
    return flex_gemm(
        F.scaled_mm,
        (hidden, w2.t(), sh, sw2),
        lambda acc: acc,
        gemm_kwargs={**SCALED_MM, "output_dtype": torch.bfloat16},
        kernel_options=options,
    )


def mlp_from_bf16(x, w1, sw1, w2, sw2, *, tuned=True):
    x, sx = quantize_mxfp8(x)
    return mlp(x, sx, w1, sw1, w2, sw2, tuned=tuned)
```    

This is mostly tests; and just wiring up the new mainloop. There is one annoying detail where we have to spoof out the actually mm_op because hops cant handle exact strides, I am working on resolving here: https://github.com/pytorch/pytorch/pull/196995

### Agent Note
Add block-scaled `F.scaled_mm` as an extension of the EpiMod cutover. The FlexGEMM HOP carries a
stride-neutral `aten.mm` surrogate because ProxyTensor cannot accept the exact-stride
`aten._scaled_mm_v2` overload as a higher-order input; the captured body still contains the real scaled-mm
node, and the lowering resolves that node once through `torch.fx.operator_schemas.normalize_function`
into a `QuackBlockScaledContract` (format, GEMM inputs, optional NVFP4 TensorWise global scales) from a
small recipe table. Matching BlockWise1x32 MXFP8 and BlockWise1x16 NVFP4 recipes are supported;
anything else (other recipes, bias, custom contraction, fast accumulation, mismatched swizzles or dtypes)
takes ordinary Inductor lowering instead of partially entering the QUACK path. Direct
`aten._scaled_mm_v2` calls are rejected with a request to use `F.scaled_mm`.

Template and runtime carry one `FlexGemmEpilogueBlockScaledConfig` and a single `blockscaled_format`
kwarg; the block scales are viewed into QuACK's blocked layout and passed to its existing block-scaled
main loop. The feature composes with the grouped-main, local-reduction, quantized-output,
captured-tensor, and autotuning machinery of the dense cutover; no scaled-mm-specific kernel or EpiOp is
added to PyTorch.

This change was prepared with assistance from an AI coding tool.

```bash
python -m pytest test/inductor/test_flex_gemm.py -q -k "scaled or nvfp4"
python -m pytest test/inductor/test_flex_gemm.py test/python_native/test_quack_vendor.py test/inductor/test_symmetric_mm.py -q
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/kernel/flex_gemm/lowering.py [hunk 1]
```diff
@@ -350,24 +410,36 @@ def lower_quack_flex_gemm(gemm_op, subgraph, args, gemm_kwargs, kernel_options):
...
         gemm_arg = placeholder_args[arg] if isinstance(arg, torch.fx.Node) else arg
         if not isinstance(gemm_arg, TensorBox):
             raise NotImplementedError("FlexGEMM lowering expects tensor GEMM operands")
         gemm_args.append(gemm_arg)
-    epilogue_arg_placeholders = flex_gemm_epilogue_arg_placeholders(
-        subgraph.graph_module, gemm_fx_node
+    epilogue_arg_placeholders = (
+        *mainloop_scale_nodes,
+        *flex_gemm_epilogue_arg_placeholders(subgraph.graph_module, gemm_fx_node),
     )
```
Comment (eellison) [inline torch/_inductor/kernel/flex_gemm/lowering.py:443]: > **AI-assisted review (Codex): [P2] Shared NVFP4 global scales break autotuning**
>
> Using the same tensor for both global scales creates duplicate kernel parameters, but in-process autotuning deduplicates its input tensors. With `tuned=True` and multiple candidates, compilation fails with `missing 1 required positional argument: 'out_ptr0'`. Distinct tensors with identical values work. Preserve argument positions during benchmarking and add a tuned shared-scale test.
>
> Reproduced on B200 using PR Python sources with a compatibility harness for the older installed native binary.

File: torch/_inductor/kernel/flex_gemm/fx_cutedsl_codegen.py [hunk 2]
```diff
@@ -786,6 +788,8 @@ def initial_env_for_params(self, params: list[str]) -> dict[torch.fx.Node, Any]:
         if self.alpha != 1:
             params.append("alpha")
             gemm_value = "(acc * alpha)"
+        for name in self.operand_names[: self.mainloop_scale_count]:
+            gemm_value = f"({gemm_value} * {name})"
```
Comment (eellison) [inline torch/_inductor/kernel/flex_gemm/fx_cutedsl_codegen.py]: > **AI-assisted review (Codex): [P2] Global-scale multiplication can turn finite results into infinity**
>
> The epilogue computes `(acc * scale_a) * scale_b`; native combines the scales first. With unit NVFP4 inputs and unit block scales, K=256, and global scales `1e37` and `1e-37`, eager returns **256** while compiled returns **inf**. Compute `acc * (scale_a * scale_b)` to avoid this intermediate overflow.
>
> Reproduced on B200 using PR Python sources with a compatibility harness for the older installed native binary.

**Review by eellison — APPROVED** [review]
